### PR TITLE
feat: Add file excerpt support and review quality improvements

### PR DESCRIPTION
**Pull request by finxo**
# Pull Request

## 📝 Summary

This PR improves code review quality by adding file excerpt support for unanchored findings (so reviewers always have context even when line anchoring fails), introducing a `prune_worktrees` operation to fix stale Git worktree metadata that blocked path reuse, and adding a GitHub diff validation gate alongside PR intent extraction and checklist description caps to reduce review noise.

## 🔧 Changes Made

- **File excerpt fallback**: When a code review finding cannot be anchored to a specific line, a relevant file excerpt is now included so the finding still has meaningful context
- **`prune_worktrees` operation**: Added `git worktree prune` support to `WorktreeService`, `GitClient`, and the client API — clears stale `.git/worktrees` metadata that caused "already registered" / "not a working tree" errors when reusing a worktree path
- **Duplicate snippet false positive fix**: Resolved an incorrect duplicate-detection hit in the line anchor resolver that was suppressing valid findings
- **GitHub diff validation gate**: Prevents review from proceeding when the diff is missing or invalid, improving overall review quality
- **PR intent extraction**: Extracts the intent/goal of the PR to better guide checklist generation and review focus
- **Checklist description caps**: Limits checklist item description length to reduce verbosity and keep reviews actionable
- **Parser hardening**: Added edge-case tests for the diff context manager to guard against regressions in context extraction

## 🧪 Testing

- [x] Unit tests added/updated (`poetry run pytest`)
- [ ] All tests passing (`make test`)
- [ ] Manual testing with `titan-dev`

New tests cover:
- `WorktreeService.prune_worktrees()` — happy path (verifies `git worktree prune` is called) and error path (verifies `WORKTREE_PRUNE_ERROR` code is returned)
- `resolve_file_read_access` — worktree trust rules and SHA verification logic
- Diff context manager edge cases to harden the parser against malformed input
- Default checklist rendering with descriptions to catch display regressions

## 📊 Logs

- [ ] No new log events

## ✅ Checklist
- [ ] Self-review done
- [ ] Follows the project's [logging rules](.claude/docs/logging.md) (no secrets, no content in logs)
- [ ] New and existing tests pass
- [x] Documentation updated if needed
- [x] Plugin documentation updated when plugin functions or parameters changed (`Plugins > Git Plugin`, `GitHub Plugin`, `Jira Plugin`)

### CODE REVIEW

File: plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py [hunk 1]
```diff
@@ -187,6 +201,172 @@ def get_hunk_for_old_line(self, path: str, line: int) -> Optional[ParsedHunk]:
...
+        if self._github_parsed is not None:
+            file_diff = self._github_parsed.files.get(path)
+            if file_diff is not None and not file_diff.hunks_consistent:
+                logger.warning(
+                    "get_publishable_lines: path=%s GitHub diff failed line-count self-check "
+                    "→ falling back to added lines",
+                    path,
+                )
+            else:
+                lines = file_diff.valid_review_lines if file_diff else frozenset()
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py]: When a GitHub diff is attached but this `path` has no entry in it (`file_diff is None`), this returns an empty frozenset instead of falling back to `context_file.added_lines`. That contradicts the documented precedence ("never fails, only narrows" → fallback 2: added lines). Since the attached diff can be assembled from files-API `patch` sections — which GitHub omits for large files — a large changed file would silently lose all inline comment placement. Consider treating `file_diff is None` the same as a self-check failure: log and fall through to the added-lines fallback, returning empty only for genuinely un-commentable files (e.g. binary).

File: plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py [hunk 2]
```diff
@@ -25,6 +26,71 @@
...
+        )
+
+    if checkout_sha != head_sha:
+        return FileReadAccess(
+            False,
+            "none",
+            f"checkout is at {checkout_sha[:8]}, PR head is {head_sha[:8]}",
+        )
+
+    if checkout_dirty:
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py]: When `has_uncommitted_changes()` fails, the caller passes `checkout_dirty=None`, and `if checkout_dirty:` treats that unknown state as clean — so the checkout gets trusted (`FileReadAccess(True, "checkout", ...)`) even though dirtiness was never verified. That contradicts the docstring's guarantee ("trusted only when it is provably at the head commit with nothing modified on top") and can mix uncommitted local edits into the review. The SHA check already fails safe when unverifiable; the dirty check should too, e.g.:

```python
if checkout_dirty or checkout_dirty is None:
    return FileReadAccess(
        False, "none",
        "checkout is at the PR head but has uncommitted changes" if checkout_dirty
        else "checkout is at the PR head but its dirty state could not be verified",
    )
```

File: plugins/titan-plugin-github/titan_plugin_github/operations/prompt_formatting_operations.py [hunk 3]
```diff
@@ -1,9 +1,66 @@
...
 
 import json
+import re
 
 from ..models.review_models import CommentContextEntry
 
+_HTML_COMMENT_RE = re.compile(r"<!--.*?-->", re.DOTALL)
+_MD_IMAGE_RE = re.compile(r"!\[[^\]]*\]\([^)]*\)")
+_CHECKBOX_LINE_RE = re.compile(r"^\s*[-*]\s*\[[ xX]\]")
+_BARE_LINK_LINE_RE = re.compile(r"^\s*\[?!?\[?[^\]]*\]?\(?https?://\S+\)?\s*$")
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/operations/prompt_formatting_operations.py]: This regex matches more than bare-link lines: `[^\]]*` lets any bracket-free prose precede the URL, so a line like `This PR adds retry logic, see https://jira.example.com/TIC-42` is dropped entirely — even though it may be the only real intent sentence in the description. Consider anchoring the pre-URL portion to link/image syntax only (e.g. `^\s*!?\[[^\]]*\]\(https?://\S+\)\s*$|^\s*https?://\S+\s*$`) so prose ending in a URL survives the trim.

File: plugins/titan-plugin-github/titan_plugin_github/operations/worktree_operations.py [hunk 4]
```diff
@@ -82,6 +67,51 @@ def setup_worktree(
...
+        full_worktree_path: Absolute path on disk; when empty, the directory
+                            removal step is skipped
+    """
+    try:
+        git_client.remove_worktree(worktree_path, force=True)
+    except Exception:
+        pass
+
+    try:
+        git_client.prune_worktrees()
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/operations/worktree_operations.py]: The prune step runs before the `rmtree` step, but `git worktree prune` only clears metadata for worktrees whose directory is *missing*. If `remove_worktree` fails while the directory still exists (e.g. a locked worktree — `git worktree remove --force` refuses those unless force is given twice), prune is a no-op here, then `rmtree` deletes the directory and leaves the stale registration behind — so the follow-up `create_worktree` fails with "missing but already registered worktree", the exact state this helper is meant to clear. Consider running `prune_worktrees()` after the directory removal (or both before and after) so metadata orphaned by the `rmtree` step is also cleared.

File: plugins/titan-plugin-github/titan_plugin_github/widgets/comment_view.py [hunk 5]
```diff
@@ -211,24 +219,42 @@ def from_suggestion(cls, suggestion: Any) -> "CommentView":
...
-            line=action.resolved_line or action.line,
+            # For a new comment, only the resolved line is a real position in this PR's
+            # diff. Falling back to the raw AI line would render a code block around a
+            # line nothing verified — extract_diff_hunk_for_action already returns no hunk
+            # in that case, so the block would be empty or, worse, from elsewhere.
+            line=action.resolved_line if is_new_comment else (action.resolved_line or action.line),
             diff_hunk=diff_hunk,
             severity=action.severity,
             line_label=_build_action_line_label(action),
+            file_excerpt=file_excerpt,
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/widgets/comment_view.py]: The excerpt path can trigger for `REPLY_TO_THREAD` actions, not just unanchored new comments. `build_thread_actions` stores reply actions under `review_action_proposals` and reuses `validate_review_actions`, where `extract_file_excerpt_for_action` gates only on `resolved_line is None` + `path`/`line` — a reply to an outdated thread satisfies all three. It then renders with the caption "pre-existing code around line N (unchanged by this PR)", which is wrong for a thread on changed-but-outdated code, and centres the excerpt on the thread's line number, which may not match the current file. Consider gating the excerpt to new comments, e.g. `file_excerpt=file_excerpt if is_new_comment else None` here, or checking `action.action_type == ReviewActionType.NEW_COMMENT` inside `extract_file_excerpt_for_action`.

File: plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py [hunk 6]
```diff
@@ -187,6 +201,172 @@ def get_hunk_for_old_line(self, path: str, line: int) -> Optional[ParsedHunk]:
...
+
+        try:
+            content = self._content_provider(path)
+        except Exception as e:
+            # A provider reads from a worktree or the network; neither should be able to
+            # break comment rendering.
+            logger.debug("get_file_content: provider failed for %s: %s", path, e)
+            content = None
+
+        self._content_cache[path] = content
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py]: A provider exception is swallowed at debug level and then cached as `None`, so a transient failure permanently disables excerpts for that path for the manager's lifetime (which `get_or_create_diff_manager` extends across the session). Consider only caching successful lookups (or genuine "file not found" `None` returns) and letting exceptions retry on the next call — and logging the failure at warning level, since it degrades the excerpt feature this PR introduces.

File: plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py [hunk 7]
```diff
@@ -118,11 +197,21 @@ def build_review_context_package(
...
+    When ``allow_file_reads`` is False, no file is read from ``cwd``: every file falls
+    back to its diff hunks. Callers set this when the content on disk cannot be proven to
+    be the PR's head revision — see ``resolve_file_read_access``.
+    """
     manager = diff_manager or get_or_create_diff_manager(diff)
     applicable_ids = set(plan.review_axes)
     checklist_applicable = [item for item in checklist if item.id in applicable_ids] or checklist[:2]
-    related_files = resolve_context_requests(plan.extra_context_requests[:1], cwd)
+    related_files = resolve_context_requests(
+        plan.extra_context_requests[:1], cwd, allow_file_reads=allow_file_reads
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py:226]: The validator allows multiple `extra_context_requests` and the plan summary reports all of them, but only the first is resolved here (`[:1]`) — the rest are dropped silently. If capping to one is intentional, either enforce that cap in `ReviewPlanValidator` / the plan prompt so the plan and the prompt agree, or at least log the dropped requests so the truncation is observable.

File: plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py [hunk 8]
```diff
@@ -751,6 +758,9 @@ def fetch_pr_review_bundle(ctx: WorkflowContext) -> WorkflowResult:
             match patches_result:
                 case ClientSuccess(data=patches_diff) if patches_diff:
                     diff = patches_diff
+                    # Files-API patches ARE GitHub's diff hunks — valid as the
+                    # publishable-lines source.
+                    diff_is_github_source = True
                 case _:
                     ctx.textual.end_step("error")
                     return Error("Could not fetch file patches for large PR")
```
Comment (finxo) [inline plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py:770]: This `case _:` swallows the ClientError message from `get_pr_file_patches` (and also silently covers an empty ClientSuccess). Other failure branches in this step include the underlying `error_message` in both the UI text and the Error — doing the same here would make large-PR failures diagnosable: `case ClientError(error_message=err): return Error(f"Could not fetch file patches: {err}")`.

**Review by finxo — COMMENTED** [review]
**plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py** (line 866):
In the outdated path, `target_line` is an old-file line number (`comment.original_line`), and `_build_focused_diff_from_hunk` correctly skips target lookup when `is_outdated=True` — but it still passes `target_line` into `_rebuild_diff`, which compares it against `new_num`. A new-file line that coincidentally shares the number gets the `◄` marker, pointing the reader at the wrong line. Suggest passing `None` as the marker target when `is_outdated` (or comparing against `old_num` in that case).

---

**plugins/titan-plugin-github/titan_plugin_github/operations/review_action_operations.py** (line 265):
This gate checks `action.line` (the AI-reported line) but the lookup below uses `action.resolved_line`. A finding with `line=null` that was anchored via a unique snippet match (`resolve_line_anchor` supports `line=None`) has a valid `resolved_line`, yet returns `None` here — and `extract_file_excerpt_for_action` won't cover it either, since it only fires when `resolved_line is None`. Such findings are shown with no code context at all. Consider gating on `resolved_line` instead: `if not diff or not action.path: return None` and keep the existing `resolved_line is None` check.

---

**plugins/titan-plugin-github/titan_plugin_github/operations/review_action_operations.py** (line 314):
This early return ignores an explicitly supplied `diff_manager`: with `diff == ""` but a valid manager, anchor resolution is skipped entirely, so every comment loses its `resolved_line` and silently falls back to the general review body. `build_review_action_payload` handles the same pair as `diff_manager or (get_or_create_diff_manager(diff) if diff else None)` — mirror that here, e.g. `if not diff and diff_manager is None: return actions`.

---

**plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py** (line 814):
A failure fetching review threads (or general comments) is silently collapsed into an empty list, and the step still returns Success. Since `review_threads` drives dedup against existing comments and thread resolution, an error here can cause duplicate comments to be re-proposed while the summary line reports "0 review thread(s)" as fact. At minimum emit `ctx.textual.warning_text(...)` like the `get_current_user` branch does, so the degraded state is visible to the user.

---

**plugins/titan-plugin-github/tests/unit/test_diff_context_manager.py** (line 531):
Two resolver-v2 branches documented in `resolve_line_anchor` aren't actually pinned by these tests:

1. **"ambiguous match, AI line valid elsewhere → the AI line wins"** — no test passes a `line` that is in `valid_review_lines` but not among the snippet matches (e.g. `line=51`, snippet=`"return None"` should return 51, not relocate to 5/53).
2. **Publishable preference** — in this test both matches (5 and 53) are added lines, so the `m not in publishable` sort key never discriminates; the test passes even if that key is removed, despite the name `prefers_publishable_then_first`. A diff where the snippet matches one context-only line and one added line (with no GitHub diff attached, so only the added line is publishable) would actually exercise it.

Given the PR is specifically about anchoring reliability, consider adding both cases — a regression in either branch (e.g. anchoring to a non-publishable line → GitHub 422) would currently slip through.

---

**plugins/titan-plugin-github/titan_plugin_github/widgets/comment_view.py** (line 312):
`badge_labels` has no entry for `ThreadSeverity.NONE`, which the constructor's type hint accepts and which is truthy (StrEnum "none"), so `badge_labels[self.severity]` would raise KeyError inside `compose()` and crash the widget. Callers currently filter NONE upstream, but the widget itself should be safe: use `badge_labels.get(self.severity)` and skip the badge when there's no label (or exclude NONE at the type level).

File: plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py (hunk 1 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py]: aa9bf584

File: plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py (hunk 2 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py]: b0799304

File: plugins/titan-plugin-github/titan_plugin_github/operations/prompt_formatting_operations.py (hunk 3 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/operations/prompt_formatting_operations.py]: 7854abdf

File: plugins/titan-plugin-github/titan_plugin_github/operations/worktree_operations.py (hunk 4 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/operations/worktree_operations.py]: 770f713d

File: plugins/titan-plugin-github/titan_plugin_github/widgets/comment_view.py (hunk 5 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/widgets/comment_view.py]: 8c0b797f

File: plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py (hunk 6 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/managers/diff_context_manager.py]: 80e89e73

File: plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py (hunk 7 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/operations/context_resolution_operations.py:226]: 807f079c

File: plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py (hunk 8 above)
Comment (finxo) [reply plugins/titan-plugin-github/titan_plugin_github/steps/code_review_steps.py:770]: 9fcb54d7
